plugins/web/opentelemetry-plugin-react-load/src/BaseOpenTelemetryComponent.ts

```diff
@@ -0,0 +1,541 @@
+    this.patch();
+  }
+
+  /**
+   * Sets the tracer for all components being instrumented
+   * @param name Name of tracer
+   * @param version Version of tracer, this is optional. When not provided it will use the latest.
+   */
+  static setTracer(name: string, version?: string): void {
+    BaseOpenTelemetryComponent._tracer = api.trace.getTracer(
+      name,
+      version ? version : VERSION
+    );
+  }
+
+  /**
+   * Sets the logger for all components being instrumented
+   * @param logger
+   */
+  static setLogger(logger: api.Logger): void {
+    BaseOpenTelemetryComponent._logger = logger;
+  }
+
+  /**
+   * Creates a new span as a child of the current parent span.
+   * If parent span is undefined, just the child is created.
+   * @param react React component currently being instrumented
+   * @param name Name of span
+   */
+  private _createSpanWithParent(
+    react: React.Component,
+    name: string
+  ): api.Span | undefined {
+    return BaseOpenTelemetryComponent._tracer.startSpan(name, {
+      attributes: this._getAttributes(react),
+      parent: this._getCreateParentSpan(react)?.span,
+    });
+  }
+
+  /**
+   * Creates a new span
+   * @param react React component currently being instrumented
+   * @param name Name of span
+   */
+  private _createSpan(
+    react: React.Component,
+    name: string
+  ): api.Span | undefined {
+    return BaseOpenTelemetryComponent._tracer.startSpan(name, {
+      attributes: this._getAttributes(react),
+    });
+  }
+
+  /**
+   * Provides instrumentation for a function
+   * @param react React component currently instrumenting.
+   * @param spanName Name to set the span of the instrumented function to.
+   * @param original Original function currently being wrapped.
+   * @parentName Name to set parent span to on error.
+   */
+  private _instrumentFunction(
+    react: React.Component,
+    spanName: string,
+    original: any
+  ) {
+    const span = this._createSpanWithParent(react, spanName);
+    let res;
+    try {
+      res = original();
+    } catch (err) {
+      if (span) {
+        span.setAttribute(AttributeNames.REACT_ERROR, err.stack);
+        span.end();
+      }
+      this._endParentSpan(react);
+      throw err;
+    }
+
+    if (span) {
+      span.end();
+    }
+    return res;
+  }
+
+  /**
+   * Ends the current parent span.
+   * @param react React component parent span belongs to.
+   */
+  private _endParentSpan(react: React.Component) {
+    const parentSpanData = this._parentSpanMap.get(react);
+    if (parentSpanData) {
+      parentSpanData.span.updateName(parentSpanData.name);
+      parentSpanData.span.end();
+      this._parentSpanMap.delete(react);
+    }
+  }
+
+  /**
+   * Returns attributes object for spans
+   * @param react React component currently being instrumented
+   **/
+  private _getAttributes(react: React.Component) {
+    let state = '';
+    try {
+      state = JSON.stringify(react.state);
```

Comment:
@opentelemetry/javascript-maintainers
Do attributes need to be non-object types? I wonder if this stringify is something that should be left to the exporters. 
Else I think the attribute value type should be changed from `unknown` to clarify whether this stringify is necessary or not.

Reply:
We already have an issue to track this on our side https://github.com/open-telemetry/opentelemetry-js/issues/1340, and a spec PR which resolved this https://github.com/open-telemetry/opentelemetry-specification/pull/722, but the simple answer is that this is not allowed:

> - The attribute key, which MUST be a non-`null` and non-empty string.
> - The attribute value, which is either:
>   - A primitive type: string, boolean, double precision floating point (IEEE 754-1985) or signed > 64 bit integer.
>   - An array of primitive type values. The array MUST be homogeneous,
>     i.e. it MUST NOT contain values of different types. For protocols that do
>     not natively support array values such values SHOULD be represented as JSON strings.